[Tankerkoenig] Changed Readme and handling of web-returns not OK.

JueBag opened the pull request:
**Changes in the Readme:**

- As reported by @cweitkamp and @hannibal29 the syntax of the example Items file had an error.
- Procedure to import hhtps certificates included
- FAQ section included

**Changes in Code:**
In case the web-request for price -data does not return status=OK al Station(s) and the Webservice go OFFLINE and the scheduled polling is stopped.

Signed-off-by: Jürgen Baginski <opus42@gmx.de>

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/WebserviceHandler.java:
```diff
@@ -139,6 +139,13 @@ public void updateStationData() {
+                    tkh.updateStatus(ThingStatus.OFFLINE);
+                }
+                // and the scheduled job gets stopped!
+                pollingJob.cancel(true);
```
Don't you want to poll in that case to check whether it works again?

JueBag replied:
IMHO  this will only be called if there was a reply from tankerkonig and this result is not "OK".  In this case there is something wrong with the request (one example is that the API-key got banned). All those reasons (are there more then one??) should be treated before new requests are sent (tankerkeonig asks for such a behaviour in its docs).
All other reasons for NO reply are treated elsewhere ( like no response due to no connection) and do not stop the polling.

martinvw replied:
So it cannot be because there is temporary failure at the servers of tankerkonig... That they return a 500 or any other errorcode. Or because they had some scheduled downtime and returned a non-json response stating their maintenance.

JueBag replied:
Good Point!
I should only treat specific json-responses from tankerkoenig that way. If the json-field "OK" is true, the message should be treated normally. If this field is false, the polling should be stopped. 
In all other failure cases the polling should continue!
Thanks for that second opinion! I'll change that this evening.

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/WebserviceHandler.java:
```diff
@@ -139,6 +139,13 @@ public void updateStationData() {
+                // if the Bridge goes OFFLINE, all Stations will go OFFLINE as well
+                for (Thing thing : getThing().getThings()) {
+                    StationHandler tkh = (StationHandler) thing.getHandler();
+                    tkh.updateStatus(ThingStatus.OFFLINE);
```
Does ESH have no smarter way to do this, will they not already be offline if the bridge is offline, or is there maybe a hook which gets called when the bridge changes status?

JueBag replied:
I might be wrong, but for me looked as if only the bridge goes OFF LINE with the current code.
I do not know of any other "smart"way to do this, but I'm not that much about ESH.

martinvw replied:
I will try to check this

JueBag replied:
You were correct, ESH does have a better way! I didn't see that a bridge has an event on which a thing can react with the method "ThingHandler.bridgeStatusChanged(ThingStatusInfo)".
I'm on it rigth now.

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/StationHandler.java:
```diff
@@ -108,6 +108,11 @@ public void dispose() {
     }
 
+    @Override
+    public void updateStatus(ThingStatus status) {
```
Why this, what is the default implementation, what went wrong?

JueBag replied:
As far as I understand that method can't be called from the outside ( from a thing to update the bridge [as in the unchanged code] and vice versa.

martinvw commented:
Thanks for the updates! I added some remarks.

JueBag commented:
Thanks for your review. I'm so for the several commits, but I was again unable to put them all in one.

The reason to have more then one commit was , in Eclipse the Readme was showing differently ( for example: no space needed between "#" and the headline in Eclipse).

JueBag commented:
I changed the code as discribed in the commit. 
The  routine "updateStatus" in the StationHandler is deleted because it isn't needed anymore. However, in the WebserviceHandler this routine is still used because the Webservice (bridge) should go ONLINE if the Station(thing) does its first positively completed web-request. I don't know any other way to set the bidge to ONLINE!

On further request:
I seem to be unable to understand correctly how to put several commits into a single one (squash?). I'm using  SourceTree to handle the commits. What is the correct way to do that?
At the end of my first contribution Kai had to do the "git-Magic", I'd like to be able to do that by myself!
Could anybody give me a hint / link .... THANKS

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/StationHandler.java:
```diff
@@ -108,6 +109,17 @@ public void dispose() {
 
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        logger.debug("Bridge Status updated to {} for device: {}", bridgeStatusInfo.getStatus().toString(),
```
You don't have to call toString it will be called automatically, but only when the logging-level actually applies.

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/StationHandler.java:
```diff
@@ -108,6 +109,17 @@ public void dispose() {
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        logger.debug("Bridge Status updated to {} for device: {}", bridgeStatusInfo.getStatus().toString(),
+                getThing().getUID().toString());
+        if (bridgeStatusInfo.getStatus().equals(ThingStatus.ONLINE)) {
```
It is an enum you can compare with == maybe even != so that you get rid of the empty if and the else

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/WebserviceHandler.java:
```diff
@@ -136,9 +136,21 @@ public void updateStationData() {
+                // if the Bridge goes OFFLINE, all connected Stations will go OFFLINE as well.
+                // The bridge reports its statusUpdate and the things react using the bridgeStatusChanged-Method!
+                // Only if the message is NOT "No valid response from the web-request!" the scheduled job gets stopped!
+                if (result.getMessage() != "No valid response from the web-request!") {
```
Strings should be compared with equals

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/WebserviceHandler.java:
```diff
@@ -136,9 +136,21 @@ public void updateStationData() {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
-                        "Empty return or no internet connection");
+                // two possibel reasons for result.isOK=false
+                // A-tankerkoenig returns false on a web-request
```
Thanks for the clear explanation here

JueBag replied:
Actually, I did explain it more to myself ;-)

martinvw commented:
i always end up here, which works for me:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/handler/WebserviceHandler.java:
```diff
@@ -136,9 +136,21 @@ public void updateStationData() {
+                // if the Bridge goes OFFLINE, all connected Stations will go OFFLINE as well.
+                // The bridge reports its statusUpdate and the things react using the bridgeStatusChanged-Method!
+                // Only if the message is NOT "No valid response from the web-request!" the scheduled job gets stopped!
+                if (!result.getMessage().equals(TankerkoenigBindingConstants.NO_VALID_RESPONSE)) {
```
Even better

martinvw commented on addons/binding/org.openhab.binding.tankerkoenig/src/main/java/org/openhab/binding/tankerkoenig/internal/config/TankerkoenigDetailResult.java:
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2010-2017 by the respective copyright holders.
+TankerkoenigBindingConstants.NO_VALID_RESPONSE * Copyright (c) 2010-2017 by the respective copyright holders.
```
This seems a little lost

JueBag replied:
Whoops!

JueBag commented:
Thanks for the link!
The picture of the situation in link doesn't match mine. 
![sourcetree](https://user-images.githubusercontent.com/18365158/27926484-e6eef60a-6288-11e7-95c8-2b4132833338.jpg)
If I understand correctly in the situation posted I should use "git rebase -i HEAD~10"
(that would include the revert).

JueBag commented:
Tried it myself.
Got this message:
```
D:\Eclipse2.2\openhab2-master\git\openhab2-addons\addons\binding\org.openhab.binding.tankerkoenig>git rebase -i HEAD~10
[detached HEAD 1ffe48709] Changes in the Readme: -As reported by @cweitkamp and @hannibal29 the syntax of the example Items file had an error. -Procedure to import hhtps certificates included -FAQ section included Changes in Code: Changed the handling of the return of a web-request. In case a valid json-response is recieved and the "OK" field is set to FALSE Webservice and Station(s) go OFFLINE, price-polling gets stopped. In case of no return, empty return etc.... Webservice and Station(s) go OFFLINE but the polling is not stopped.
 Date: Tue Jul 4 21:20:13 2017 +0200
 8 files changed, 134 insertions(+), 38 deletions(-)
Successfully rebased and updated refs/heads/master.
```
But it doesn't show in SourceTree nor in here????

martinvw commented:
You should get a view where you choose to pick, squash or even ignore the commits and you should pick the first one and ignore the others. If everything went fine you should be able to do a forced push after that.  SourceTree supports that but you have to enable it first in the options dialog.

JueBag commented:
I did the rebase in the command window of windows. I got those views, or better text editors (I had to search for the saving command of vi!) and made the desired changes to the commits and next to the messages.
After looking a bit more in SourceTree and selecting "Show branches from Remote" (my translation!) it looks in thee as if it worked. The message from the command window said that also.  
Only in here I don't see that change.

martinvw commented:
Have you and / or others tested this newest version? Should it be tested before merging? What do you think (wdyt)?

JueBag commented:
After changing the way, the web-response is handled, I did test  it using false IDS for the API-key and for a Stations and with a lost Internet connection  in order to see what response would be received. 
After the minor changes from yesterday, I didn't do it again.

Sorry for my "stepp-wise" understanding of this "Git-magic", especially the last part  (forced push) wasn't mentioned in the link.

martinvw commented:
> Sorry for my "stepp-wise" understanding of this "Git-magic", especially the last part (forced push) wasn't mentioned in the link.

No problem at all.
